### ISSUE TITLE
ScriptEvaluationError when a property includes a dollar sign ($)

### ISSUE THREAD

**Issue by davidsblog** [evo-terren/Gremlin.Net.CosmosDb#23]
Hi,

I noticed that if you try to use a string property which contains a $ char then a ScriptEvaluationError is thrown.  Here is some sample code to demonstrate it:

```
internal static async Task Main()
{
    using (var graphClient = new GraphClient("your-gremlin-host-name", "your-db-name", "your-graph-name", "your-access-key"))
    {
        var g = graphClient.CreateTraversalSource();

        //add vertices/edges using strongly-typed objects
        var personV = new PersonVertex
        {
            Ages = new[] { 4, 6, 23 },
            Id = "person-12345",
            Label = "some-label",
            Name = "my $ name"
        };
        var purchasedE = new PersonPurchasedProductEdge
        {
            Id = "person-12345_purchased_product-12345"
        };
        var productV = new ProductVertex
        {
            Id = "product-12345"
        };
        var test = g.AddV(personV).As("person")
                    .AddV(productV).As("product")
                    .AddE(purchasedE).From("person").To("product");

        Console.WriteLine(test.ToGremlinQuery());
        var response = await graphClient.QueryAsync(test);
    }

    Console.WriteLine();
    Console.WriteLine("All done...");
    Console.ReadKey();
}
```

You'll see that I have just taken the sample code and changed the **Name** property to include a dollar sign.  The error message is:

```
ScriptEvaluationError: 

ActivityId : xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
ExceptionType : GraphCompileException
ExceptionMessage :
	Gremlin Query Compilation Error: Script compile error: Missing ')' @ line 1, column 129.
Source : Microsoft.Azure.Graphs
	GremlinRequestId : xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
	Context : graphcompute
	Scope : graphparse-translate-outer
	GraphStatusCode : ScriptEvaluationError
	HResult : 0x80131500
```

Thanks, David

**Comment by davidsblog** [comment]
This may be an underlying problem in Microsoft's implementation.  I've noticed the problem can also be reproduced if you execute the following query in the Azure Portal (Data Explorer):

`g.addV("person").property("name","Chuck Norri$")`

... which throws an exception in the portal too.  Although it works if I try this:

`g.addV("person").property("name","Chuck Norri\u0024")`

But it would be nice if something could be done in this library to automatically convert '$' to '\u0024' so that any string values with a dollar sign don't throw exceptions.

**Comment by BenjaBobs** [comment]
This is because CosmosDB does not support `GStrings` (http://groovy-lang.org/syntax.html#all-strings).
I haven't tested what other features are not supported yet, but prepending with a `\` appears to do the trick in terms of escaping characters.

**Comment by davidsblog** [comment]
Many thanks @BenjaBobs, you're absolutely right, prepending a `\` also works fine.  Thank-you for explaining about `GString` support.  Do you think that anything can be done in this project to help, since any string could contain a dollar sign? The following _dirty hack_ in `GraphClient.cs` solves the problem for me, but I don't like it:

```
public Task<IReadOnlyCollection<JToken>> QueryAsync(string gremlinQuery)
{
    if (gremlinQuery == null)
        throw new ArgumentNullException(nameof(gremlinQuery));

    gremlinQuery = gremlinQuery.Replace("$", "\\$"); // dirty hack

    return _gremlinClient.SubmitAsync<JToken>(gremlinQuery);
}
```

Thanks for your help!  I'm finding this project very useful, much better than having graph queries as hard-coded strings...

**Comment by BenjaBobs** [comment]
I'm using the exact same dirty hack, but ideally CosmosDB would support full Gremlin, or this project would have some sort of handling of `GString` and other unsupported Groovy features.
I haven't come up with a good way to find out what else CosmosDB doesn't support yet, but at least we have the list of keywords and special characters for Groovy.

**Comment by evo-terren** [comment]
I apologize for taking so long to chime in on this one. I've been thinking about it a bit. I think escaping the special characters would be the best thing to do going forward. I do have some hesitancy with not allowing users to utilize the `GString` and all it supports, but I wonder if @BenjaBobs is right. Maybe the best way to make it clear is to have some kind of `GString` object that accepts a raw string and doesn't escape it at all. I can't imagine very many people needing such a feature, but it would be simple to implement and would be there for the advanced users who need it. In the end, the script evaluation error you received will be a shock to most and that can be avoided by doing the proper escaping.

Would either of you be able to compile a list of the characters that need escaping? Is it simply the quotes and dollar sign characters?

**Comment by BenjaBobs** [comment]
I did a bunch of tests. Here are the results: https://pastebin.com/yWpb0q0Y

**Comment by davidsblog** [comment]
Thanks @BenjaBobs and @evo-terren!

The only char that I've had a problem with is the `$` so far.  I threw together some nasty code to try a few other chars, like this:

```
var test = "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";
for (int c = 0; c < test.Length; c++)
{
    try
    {
        var personV = new PersonVertex
        {
            Name = $"index #{c} testing for this char {test[c]} code={(int)test[c]} thanks"
        };
        var add = g.AddV<PersonVertex>(personV);
        var added = (await graphClient.QueryAsync(add)).FirstOrDefault();

        var check = g.V<PersonVertex>(added.Id);
        var checkd = (await graphClient.QueryAsync(check)).FirstOrDefault();

        if (checkd.Name != personV.Name)
        {
            Console.WriteLine($"No match: {c} {test[c]} ({(int)test[c]})");
        }
    }
    catch (Exception)
    {
        Console.WriteLine($"Error: {c} {test[c]} ({(int)test[c]})");
    }   
}
```

**NOTE:** my _dirty hack_ was in place to fix the `$` problem, and if I remove it, the only char that fails is the `$` symbol. So, the result ... using `g.V().values("Name")` was this:

```
[
  "index #0 testing for this char ! code=33 thanks",
  "index #1 testing for this char \" code=34 thanks",
  "index #2 testing for this char # code=35 thanks",
  "index #3 testing for this char $ code=36 thanks",
  "index #4 testing for this char % code=37 thanks",
  "index #5 testing for this char & code=38 thanks",
  "index #6 testing for this char ' code=39 thanks",
  "index #7 testing for this char ( code=40 thanks",
  "index #8 testing for this char ) code=41 thanks",
  "index #9 testing for this char * code=42 thanks",
  "index #10 testing for this char + code=43 thanks",
  "index #11 testing for this char , code=44 thanks",
  "index #12 testing for this char - code=45 thanks",
  "index #13 testing for this char . code=46 thanks",
  "index #14 testing for this char / code=47 thanks",
  "index #15 testing for this char : code=58 thanks",
  "index #16 testing for this char ; code=59 thanks",
  "index #17 testing for this char < code=60 thanks",
  "index #18 testing for this char = code=61 thanks",
  "index #19 testing for this char > code=62 thanks",
  "index #20 testing for this char ? code=63 thanks",
  "index #21 testing for this char @ code=64 thanks",
  "index #22 testing for this char [ code=91 thanks",
  "index #23 testing for this char \\ code=92 thanks",
  "index #24 testing for this char ] code=93 thanks",
  "index #25 testing for this char ^ code=94 thanks",
  "index #26 testing for this char _ code=95 thanks",
  "index #27 testing for this char ` code=96 thanks",
  "index #28 testing for this char { code=123 thanks",
  "index #29 testing for this char | code=124 thanks",
  "index #30 testing for this char } code=125 thanks",
  "index #31 testing for this char ~ code=126 thanks"
]
```

You have to remember that's a JSON response, but it all looks OK to me. I hope that helps.

**Comment by evo-terren** [comment]
I also found that the dollar sign seemed to be the only offender. I just released a new version [v0.3.3.2016-alpha](https://github.com/evo-terren/Gremlin.Net.CosmosDb/releases/tag/v0.3.3.2016-alpha) which should address this serialization issue.

I also created a `GroovyString` object which can be used for advanced usage for anyone who would like to utilize more features with the `groovy.lang.GString`. Hopefully this will work out for everyone.

Let me know if you have any more issues.
